[OSDOCS#12841] Observability metrics correlation for ShiftStack (#86741)
openshift/openshift-docs: modules/monitoring-configuring-shiftstack-scraping.adoc

[id="monitoring-configuring-shiftstack-scraping_{context}"]
= Collecting cluster metrics from the federation endpoint

You can employ the federation endpoint of your {product-title} cluster to make metrics available to a {rhoso-first} cluster to practice pull-based monitoring.

.Prerequisites

- You have administrative access to {rhoso} and the tenant cluster that is running on it.
- Telemetry is enabled in the {rhoso} environment.
- The Cluster Observability Operator is installed on your cluster.
- The monitoring stack for your cluster is configured.
- Your cluster has its federation endpoint exposed.

.Procedure

. Connect to your cluster by using a username and password; do not log in by using a `kubeconfig` file that was generated by the installation program.

. To retrieve a token from the {product-title} cluster, run the following command on it:
+
[source,terminal]
----
$ oc whoami -t
----

. Make the token available as a secret in the `openstack` namespace in the {rhoso} management cluster by running the following command:
+
[source,terminal]
----
$ oc -n openstack create secret generic ocp-federated --from-literal=token=<the_token_fetched_previously>
----

> . To get the Prometheus federation route URL from your {product-title} cluster, enter the following command:

Author: ```suggestion
. To get the Prometheus federation route URL from your {product-title} cluster, run the following command:
```

// Module included in the following assemblies:
//
// * observability/monitoring/shiftstack-prometheus-configuration.adoc

:_mod-docs-content-type: PROCEDURE
[id="monitoring-configuring-shiftstack-scraping_{context}"]
= Collecting cluster metrics from the federation endpoint

You can employ the federation endpoint of your {product-title} cluster to make metrics available to a {rhoso-first} cluster to practice pull-based monitoring.

.Prerequisites

- You have administrative access to {rhoso} and the tenant cluster that is running on it.
- Telemetry is enabled in the {rhoso} environment.
- The Cluster Observability Operator is installed on your cluster.
- The monitoring stack for your cluster is configured.
- Your cluster has its federation endpoint exposed.

.Procedure

. Connect to your cluster by using a username and password; do not log in by using a `kubeconfig` file that was generated by the installation program.

> . To retrieve a token from the management cluster, run the following command on it:

Reviewer: ```suggestion
. To retrieve a token from the {product-title} cluster, run the following command on it:
```

+
[source,terminal]
----
$ oc whoami -t
----

> . Make the token available as a secret in the `openstack` namespace by running the following command:

Reviewer: ```suggestion
. Make the token available as a secret in the `openstack` namespace in the {rhoso} management cluster by running the following command:
```

+
[source,terminal]
----
$ oc -n openstack create secret generic ocp-federated --from-literal=token=<the_token_fetched_previously>
----
+

> Note that `ocp-federated` is the name of the secret in the previous `ScrapeConfig` manifest.

Reviewer: ```suggestion
```

> . To get the Prometheus federation route URL from your cluster, enter the following command:

Reviewer: ```suggestion
. To get the Prometheus federation route URL from your {product-title} cluster, enter the following command:
```

+
[source,terminal]
----
$ oc -n openshift-monitoring get route prometheus-k8s-federate -ojsonpath={'.status.ingress[].host'}
----

. Write a manifest for a scrape configuration and save it as a file called `cluster-scrape-config.yaml`. As an example:
+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  labels:
    service: metricStorage
  name: sos1-federated
  namespace: openstack
spec:
  params:
    'match[]':
    - '{__name__=~"kube_node_info|kube_persistentvolume_info|cluster:master_nodes"}' # <1>
  metricsPath: '/federate'
  authorization:
    type: Bearer
    credentials:
      name: ocp-federated # <2>
      key: token
  scheme: HTTPS # or HTTP
  scrapeInterval: 30s # <3>
  staticConfigs:
  - targets:
    - prometheus-k8s-federate-openshift-monitoring.apps.openshift.example # <4>
----
<1> Add metrics here. In this example, only the metrics `kube_node_info`, `kube_persistentvolume_info`, and `cluster:master_nodes` are requested.

> <2> Note the name of the secret for later.

Reviewer: ```suggestion
<2> Insert the previously generated secret name here.
```

<3> Limit scraping to fewer than 1000 samples for each request with a maximum frequency of once every 30 seconds.
<4> Insert the URL you fetched previously here. If the endpoint is HTTPS and uses a custom certificate authority, add a `tlsConfig` section after it.

> . While connected to the management cluster, apply the manifest by running the following command:

Reviewer: ```suggestion
. While connected to the {rhoso} management cluster, apply the manifest by running the following command:
```

// Module included in the following assemblies:
//
// * observability/monitoring/shiftstack-prometheus-configuration.adoc

:_mod-docs-content-type: PROCEDURE
[id="monitoring-configuring-shiftstack-scraping_{context}"]
= Collecting cluster metrics from the federation endpoint

> You can open the federation endpoint of your {product-title} cluster to make metrics available to a {rhoso-first} cluster to practice pull-based monitoring.

Reviewer 2: The federation endpoint is OpenShift's, so not influenced by the OpenStack version. However, the whole observability stack in OpenStack changed in RHOSO, so I don't expect the rest of the process to apply exactly.

> You can open the federation endpoint of your {product-title} cluster to make metrics available to a {rhoso-first} cluster to practice pull-based monitoring.

Reviewer: As far as I understand there is no need to open any federation endpoint, it's already opened so we can just query it. @pierreprinetti @mandre please correct me if I'm wrong.
```suggestion
You can make use of the federation endpoint of your {product-title} cluster to make metrics available to a {rhoso-first} cluster to practice pull-based monitoring.
```

Author: ```suggestion
You can employ the federation endpoint of your {product-title} cluster to make metrics available to a {rhoso-first} cluster to practice pull-based monitoring.
```

- The monitoring stack for your cluster is configured.
- Your cluster has its federation endpoint exposed.
// That is, https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/monitoring/accessing-third-party-monitoring-apis#monitoring-querying-metrics-by-using-the-federation-endpoint-for-prometheus_accessing-monitoring-apis-by-using-the-cli

.Procedure

. Connect to your cluster by using a username and password; do not log in by using a `kubeconfig` file that was generated by the installation program.

. To retrieve a token from the cluster, run the following command:
+
[source,terminal]
----
$ oc whoami -t
----

. To get the Prometheus federation route URL from your cluster, enter the following command:
+
[source,terminal]
----
$ oc -n openshift-monitoring get route prometheus-k8s-federate -ojsonpath={'.status.ingress[].host'}
----

. Write a manifest for a scrape configuration and save it as a file called `cluster-scrape-config.yaml`. As an example:
+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  labels:
    service: metricStorage
  name: sos1-federated
  namespace: openstack
spec:
  params:
    'match[]':

>     - '{__name__=~"kube_node_info|kube_persistentvolume_info"}' # <1>

Reviewer: Per https://github.com/openshift/installer/pull/9449
```suggestion
    - '{__name__=~"kube_node_info|kube_persistentvolume_info|cluster:master_nodes"}' # <1>
```

  metricsPath: '/federate'
  authorization:
    type: Bearer
    credentials:
      name: ocp-federated # <2>
      key: token
  scheme: HTTPS # or HTTP
  scrapeInterval: 30s # <3>
  staticConfigs:
  - targets:
    - prometheus-k8s-federate-openshift-monitoring.apps.openshift.example # <4>
----

> <1> Add metrics here. In this example, only the metrics `kube_node_info` and `kube_persistentvolume_info` are requested.

Reviewer: Per https://github.com/openshift/installer/pull/9449:
```suggestion
<1> Add metrics here. In this example, only the metrics `kube_node_info`, `kube_persistentvolume_info` and `cluster:master_nodes`  are requested.
```

Author: ```suggestion
<1> Add metrics here. In this example, only the metrics `kube_node_info`, `kube_persistentvolume_info`, and `cluster:master_nodes` are requested.
```

<2> Note the name of the secret for later.
<3> Limit scraping to fewer than 1000 samples for each request with a maximum frequency of once every 30 seconds.
<4> Insert the URL you fetched previously here. If the endpoint is HTTPS and uses a custom certificate authority, add a `tlsConfig` section after it.

. While connected to the {product-title} cluster that hosts the RHOSO control plane, apply the manifest by running the following command:
+
[source,terminal]
----
$ oc apply -f cluster-scrape-config.yaml
----

. Make the token available as a secret in the `openstack` namespace by running the following command:
+
[source,terminal]
----
$ oc -n openstack create secret generic ocp-federated --from-literal=token=<the_token_fetched_previously>
----
+

> Note that `ocp-federated` is the name of the secret in the previous `ScrapeConfig` manifest.

Reviewer: I would move this step to the third position in the list, just after the `$ oc whoami -t` step. Otherwise we would be creating the `scrapeconfig` object pointing to a secret (`ocp-federated`) that doesn't exist, which would probably fail.
As the creation of the secret needs to be done in the RHOSO's management OCP cluster I would clearly state that as there are other steps that need to be done in the shift-on-stack cluster.

----

. Write a manifest for a scrape configuration and save it as a file called `cluster-scrape-config.yaml`. As an example:
+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  labels:
    service: metricStorage
  name: sos1-federated
  namespace: openstack
spec:
  params:
    'match[]':
    - '{__name__=~"kube_node_info|kube_persistentvolume_info"}' # <1>
  metricsPath: '/federate'
  authorization:
    type: Bearer
    credentials:
      name: ocp-federated # <2>
      key: token
  scheme: HTTPS # or HTTP
  scrapeInterval: 30s # <3>
  staticConfigs:
  - targets:
    - prometheus-k8s-federate-openshift-monitoring.apps.openshift.example # <4>
----
<1> Add metrics here. In this example, only the metrics `kube_node_info` and `kube_persistentvolume_info` are requested.
<2> Note the name of the secret for later.
<3> Limit scraping to fewer than 1000 samples for each request with a maximum frequency of once every 30 seconds.
<4> Insert the URL you fetched previously here. If the endpoint is HTTPS and uses a custom certificate authority, add a `tlsConfig` section after it.

> . While connected to the {product-title} cluster that hosts the RHOSO control plane, apply the manifest by running the following command:

Author: FIXME: naming

// Module included in the following assemblies:
//
// * observability/monitoring/shiftstack-prometheus-configuration.adoc

:_mod-docs-content-type: PROCEDURE
[id="monitoring-configuring-shiftstack-scraping_{context}"]
= Collecting cluster metrics from the federation endpoint

// TODO: Is this RHOSO only?

> You can open the federation endpoint of a cluster to make metrics available to a Red Hat OpenStack Services on OpenShift (RHOSO) cluster to practice pull-based monitoring.

Reviewer 3: Do we need to make it clear this is an alternative to the above method of sending metrics to an external Prometheus server?

Author: Can't hurt.

.Prerequisites

- You have access to an external Prometheus instance.
- You have administrative access to {rh-openstack} and your cluster.
- You have certificates for secure communication with mTLS.
- Telemetry is enabled in the {rh-openstack} environment.
- The Cluster Observability Operator is installed on your cluster.

> - The monitoring stack for your cluster is configured.

Reviewer 3: I think it doesn't hurt to leave it there.

----

. Write a manifest for a scrape configuration and save it as a file called `cluster-scrape-config.yaml`. As an example:
+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  labels:
    service: metricStorage
  name: sos1-federated
  namespace: openstack
spec:
  params:
    'match[]':
    - '{__name__=~"kube_node_info|kube_persistentvolume_info"}' # <1>
  metricsPath: '/federate'
  authorization:
    type: Bearer
    credentials:
      name: ocp-federated # <2>
      key: token
  scheme: HTTPS # or HTTP
  scrapeInterval: 30s # <3>
  staticConfigs:
  - targets:
    - prometheus-k8s-federate-openshift-monitoring.apps.openshift.example # <4>
----
<1> Add metrics here. In this example, only the metrics `kube_node_info` and `kube_persistentvolume_info` are requested.
<2> Note the credential name for later.
<3> Limit scraping to fewer than 1000 samples for each request with a maximum frequency of once every 30 seconds.
<4> Insert the URL you fetched previously here. If the endpoint is HTTPS and uses a custom certificate authority, add a `tlsConfig` section after it.

> . While connected to the RHOSO cluster, apply the manifest by running the following command:

Reviewer 3: It's "while connected to the OpenShift cluster hosting the RHOSO control plane" actually, but that might be a bit too long. Not sure that everyone understand that by "RHOSO" cluster, we're in fact talking about the OpenShift cluster it's running on.

Author: I'll take verbosity over vagueness. 

```suggestion
. While connected to the {product-title} cluster that hosts the RHOSO control plane, apply the manifest by running the following command:
```

Author: @pierreprinetti You're correct! 

We don't need to use `{product-title}` at all if it complicates things. In a dumb substitution sense, it applies to OpenShift core in most [permutations](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#product-name-and-version) of the documentation. Most writers, self included, unthinkingly sub in `{product-title}` whenever we see `OpenShift` or things like it.

I need to consult with my CS, I think, as there are branding issues here as well as ones of elegance in writing. We'll probably end up with a new attribute specifically for this situation.

Author: @pierreprinetti Any thoughts on "management cluster" vs. "workload cluster"?

Reviewer 2: RHOSO users may have internalised the idea that "management cluster" is the name of the OpenShift cluster hosting RHOSO's control plane.

I find "workload cluster" trickier, because I imagine "workload" as being "the actual software that runs your business logic, as opposed to the infrastructure stack"... but the OpenShift cluster that runs on top of OpenStack is not your business logic application, it's just the last infrastructure layer under it.

That being said: if you find a better name, rest assured I will use it from now on.

+
[source,terminal]
----
$ oc apply -f cluster-scrape-config.yaml
----

. Make the token available as a secret in the `openstack` namespace by running the following command:
+
[source,terminal]
----
$ oc -n openstack create secret generic ocp-federated --from-literal=token=<the_token_fetched_previously>
----
+

> Note that `ocp-federated` is the name of the secret in the manifest YAML file.

Author: ```suggestion
Note that `ocp-federated` is the name of the secret in the previous `ScrapeConfig` manifest.
```
